pool.go

```diff
@@ -0,0 +1,124 @@
+package go_pool
+
+import (
+	"errors"
+	"runtime"
+	"sync"
+	"time"
+)
+
+var (
+	ErrWorkerClosed = errors.New("no routine is active")
+	ErrJobTimedOut  = errors.New("job request timed expired")
+
+	once    sync.Once
+	poolVar *Pool
+)
+
+// Pool is a struct which contains the list of routines
+type Pool struct {
+	WorkerFun func() Worker
+
+	ReqChan chan interface{}
+	RetChan chan interface{}
+}
+
+// Worker is an interface representing the routine agent
+type Worker interface {
+	Process(interface{}) interface{}
+}
+
+// Initialize is the public function which creates the pool
+func Initialize(nCpus int, nRoutines int, f func(interface{}) interface{}) *Pool {
+	setCpuToBeUsed(nCpus)
+	return New(nRoutines, func() Worker {
+		return &initWorker{
+			processor: f,
+		}
+	})
+}
+
+func New(n int, wFun func() Worker) *Pool {
+	once.Do(func() {
+		poolVar = &Pool{
+			WorkerFun: wFun,
+			ReqChan:   make(chan interface{}, n),
+			RetChan:   make(chan interface{}, n),
+		}
+		go poolVar.initWorkers(n)
+	})
+	return poolVar
+}
+
+func (p *Pool) initWorkers(n int) {
+	var wg sync.WaitGroup
+	for i := 0; i < n; i++ {
+		wg.Add(1)
+		go p.worker(&wg, p.WorkerFun())
+	}
+	wg.Wait()
+}
+
+func (p *Pool) worker(wg *sync.WaitGroup, worker Worker) {
+	for req := range p.ReqChan {
+		p.RetChan <- worker.Process(req)
+	}
+	wg.Done()
+}
+
+func (p *Pool) Process(reqPayload interface{}) interface{} {
+
+	p.ReqChan <- reqPayload
+
+	retPayload, retOk := <-p.RetChan
+	if !retOk {
+		panic(ErrWorkerClosed)
```

Comment:
why panic, why not return error.

Reply:
Removed Panic and throwing the proper error for the cause. Ref - [Commit 6debada](https://github.com/nitinks-dnz/go-pool/pull/1/commits/6debadacb7e0141868462ef5c953f08e99e001e6)